src/java.base/share/classes/java/lang/runtime/ExactConversionsSupport.java

```diff
@@ -0,0 +1,260 @@
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package java.lang.runtime;
+
+/**
+ * A casting conversion is considered to be exact if, after applying the conversion
+ * to the value, no exception was raised or loss of information has occured;
+ * otherwise it is considered inexact.
+ * The methods in this class provide the run-time support for primitive conversions exactness checks.
+ * These methods may be used, for example, by Java compiler implementations to implement checks
+ * for `instanceof` and pattern matching runtime implementations.
+ * See JLS section 5.5.1 for more information on exact casting conversions.
+ *
+ * @jls 5.5.1 Exact Casting Conversions
+ * @jls 15.20.2 The instanceof Operator
+ *
+ * @implNote Methods that describe a conversion check among unconditionally exact types
+ * (they would return true) are ommited. Similarly, methods that describe a conversion check
+ * which can be redirected through one of the methods in this class are omitted too (i.e.,
+ * byte to char can be redirected safely to `isIntToCharExact`, similarly `short` to `byte`,
+ * `short` to `char`, `char` to `byte` and `char` to `short`).
+ *
+ * @since 22
+ */
+public final class ExactConversionsSupport {
+
+    private ExactConversionsSupport() { }
+
+    /**
+     * Exactness method from int to byte
+     * @param n value
+     * @return  true if the passed value can be converted exactly to the target type
+     */
+    public static boolean isIntToByteExact(int n)      {return n == (int)(byte)n;}
+
+    /**
+     * Exactness method from int to short
+     * @param n value
+     * @return  true if the passed value can be converted exactly to the target type
+     */
+    public static boolean isIntToShortExact(int n)     {return n == (int)(short)n;}
+
+    /**
+     * Exactness method from int to char
+     * @param n value
+     * @return  true if the passed value can be converted exactly to the target type
+     */
+    public static boolean isIntToCharExact(int n)      {return n == (int)(char)n;}
+
+    /**
+     * Exactness method from int to float
+     * @param n value
+     * @return  true if the passed value can be converted exactly to the target type
+     *
+     * @implSpec relies on the notion of representation equivalence defined in the
+     * specification of the {@linkplain Double} class.
+     */
+    public static boolean isIntToFloatExact(int n) { return n == (int)(float)n && n != Integer.MAX_VALUE; }
```

Comment:
```suggestion
    public static boolean isIntToFloatExact(int n) {
        return n == (int)(float)n && n != Integer.MAX_VALUE;
    }
```
Nit: what about formatting this on 3 lines?

Reply:
Looks better indeed.